### PR TITLE
front: block state : display block state for selected train on map on simulation's result

### LINKED ISSUE #2826: Display block state for selected train on map on simulation's result
### Description and goal

# Description
Display block state for selected train on map on simulation's result, dynamically updated by time position of cursor on map or other charts.

Blocks states follow signals state. Only 2 blocks, before and current, along the path of selected train.

### Acceptance criteria

For the selected train **only**, display on map the block state along the path for:
- [x] current position of train in red
- [x] block before (in space) occupied block in yellow

The state and color of blocks are determined by signals & routes aspects.

## Beware of
- [x] optimize: the new dynamic layer has to be very tiny and contains only necessary informations, try to benchmark each solution considered
- [ ] performances: the position of train on map is actually very slow, investigate & search for the reason (perhaps the sprite of train orientation ?) and/or put a switch to disable what causes slow display, test & try best solutions to do that **THIS HAS TO BE DONE FIRST**

### PR DESCRIPTION
**Pull request by Akctarus**
close #2826

### CODE REVIEW

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
+import { Source } from 'react-map-gl';
+import lineSliceAlong from '@turf/line-slice-along';
+import { Feature, LineString } from 'geojson';
+import { RootState } from 'reducers';
+import OrderedLayer from 'common/Map/Layers/OrderedLayer';
+import { TrainPosition } from './types';
+
+function setFirstBlockState(
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: `FirstBlockState` is a bit meaningless, caould you choose a more explicit name ?

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  if (headDistance - tailDistance > threshold) {
+    return lineSliceAlong(geojsonPath, tailDistance, headDistance);
+  }
+  if (headDistance > threshold) {
+    return lineSliceAlong(geojsonPath, headDistance - threshold, headDistance);
+  }
+  return lineSliceAlong(geojsonPath, 0, threshold);
+}
+
+function setSecondBlockState(
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: Same as before

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          <OrderedLayer
+            id={`${point.id}-path3`}
+            type="line"
+            paint={{
+              'line-width': 3,
+              'line-color': '#ffb612',
+            }}
+            layerOrder={layerOrder}
+          />
+        </Source>
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: Please use one Source with two Layers

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  if (geojsonPath && point.headDistanceAlong && point.tailDistanceAlong) {
+    const [firstBlockState] = getFirstBlockState(geojsonPath, 1, 0);
+    const [secondBlockState] = getSecondBlockState(geojsonPath, 1, 0);
+    return (
+      <>
+        <Source type="geojson" data={firstBlockState}>
+          <OrderedLayer
+            id={`${point.id}-path2`}
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: No need for ID, let maplibre choose one for you (you can remove the line)

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            paint={{
+              'line-width': 3,
+              'line-color': '#cd0037',
+            }}
+            layerOrder={layerOrder + 1}
+          />
+        </Source>
+        <Source type="geojson" data={secondBlockState}>
+          <OrderedLayer
+            id={`${point.id}-path3`}
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: No need for ID, let maplibre choose one for you (you can remove the line)

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap.tsx [hunk 6]
```diff
@@ -504,6 +505,14 @@ const Map: FC<MapProps> = ({ setExtViewport }) => {
           />
         )}
 
+        {geojsonPath && selectedTrainHoverPosition && (
+          <RenderBlockState
+            point={selectedTrainHoverPosition}
+            geojsonPath={geojsonPath}
+            layerOrder={LAYER_GROUPS_ORDER[LAYERS.TRAIN.GROUP] - 1}
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap.tsx]: Don't choose a number, use a predefined LAYER (we can talk about this)

File: front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  layerOrder: number;
+}
+
+export default function RenderBlockState(props: RenderBlockStateProps) {
+  const { point, geojsonPath, layerOrder } = props;
+  const { selectedTrain } = useSelector((state: RootState) => state.osrdsimulation);
+  const simulation = useSelector((state: RootState) => state.osrdsimulation.simulation.present);
+  const data = simulation.trains[selectedTrain].base.route_aspects;
+
+  function getFirstBlockState(geojsonPath: Feature<LineString>, number1: number, number2: number) {
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap/RenderBlockState.tsx]: `number1` is meaningless, name your variables with explicit names of what it represents not what it contains (the type is enough !)

**Review by nicolaswurtz — CHANGES_REQUESTED** [review]

File: docker-compose-mac.yml [hunk 8]
```diff
@@ -123,4 +123,4 @@ services:
       editoast: {condition: service_healthy}
     restart: "no"
     image: alpine:latest
-    command: "true"
+    command: "true"
```
Comment (flomonster) [inline docker-compose-mac.yml]: Missing trailing line feed. Check how to configure it under mac: https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files

Comment (Akctarus) [reply docker-compose-mac.yml]: Done

**Review by nicolaswurtz — APPROVED** [review]

**Pull request merged**